portal-bridge/src/census.rs

```diff
@@ -0,0 +1,287 @@
+                    .await
+            }
+            ContentKey::State(content_key) => {
+                self.state
+                    .get_interested_enrs(content_key.content_id())
+                    .await
+            }
+            ContentKey::Beacon(content_key) => {
+                self.beacon
+                    .get_interested_enrs(content_key.content_id())
+                    .await
+            }
+        }
+    }
+}
+
+/// The network struct is responsible for maintaining a list of known peers
+/// in the given subnetwork.
+struct Network {
+    peers: HashMapDelay<[u8; 32], (Enr, U256)>,
+    client: HttpClient,
+    subnetwork: Subnetwork,
+}
+
+impl Network {
+    fn new(client: HttpClient, subnetwork: Subnetwork) -> Self {
+        Self {
+            peers: HashMapDelay::new(LIVENESS_CHECK_DELAY),
+            client,
+            subnetwork,
+        }
+    }
+
+    // We initialize a network with a random rfn lookup to get an initial view of the network
+    // and then iterate through the rfn of each peer to find new peers. Since this initialization
+    // blocks the bridge's gossip feature, there is a tradeoff between the time taken to initialize
+    // the census and the time taken to start gossiping. In the future, we might consider updating
+    // the initialization process to be considered complete after it has found ~100% of the network
+    // peers. However, since the census continues to iterate through the peers after initialization,
+    // the initialization is just to reach a critical mass of peers so that gossip can begin.
+    async fn init(&mut self) {
+        let random_enr = generate_random_remote_enr().1;
+        let Ok(initial_enrs) = self
+            .subnetwork
+            .recursive_find_nodes(&self.client, random_enr.node_id())
+            .await
+        else {
+            panic!("Failed to initialize network census");
+        };
+
+        for enr in initial_enrs {
+            self.iterate_peer_routing_table(enr.clone()).await;
+        }
+        info!(
+            "Initialized state census: found peers: {}",
+            self.peers.len()
+        );
+    }
+
+    /// Only processes an enr (iterating through its rfn) if the enr's
+    /// liveness delay has expired
+    async fn process_enr(&mut self, enr: Enr) {
+        // ping for liveliness check
+        if !self.liveness_check(enr.clone()).await {
+            return;
+        }
+        // iterate through the peer's rfn
+        self.iterate_peer_routing_table(enr).await;
+    }
+
+    // iterate peers routing table via rfn over various distances
+    async fn iterate_peer_routing_table(&mut self, enr: Enr) {
+        for distance in 245..257 {
+            let Ok(result) = self
+                .subnetwork
+                .find_nodes(&self.client, enr.clone(), vec![distance])
+                .await
+            else {
+                continue;
+            };
+            for found_enr in result {
```

Comment:
nit: is it worth doing this in parallel?

Reply:
Probably, cause it's definitely a bit slow as it is. But it's also a bit tricky to pull it off in parallel as it's currently written since you'd need to pass a mutable reference across threads... Though, there's definitely room for improvement here. But since the state network is currently small and quick to initialize, I'm just going to leave a comment saying that it should be improved once initialization becomes too slow